### PR TITLE
For #9606 - Adds defaultSettings when creating a new GeckoEngineSession

### PR DESCRIPTION
**Pull request by codrut-topliceanu**
This fixes the issue where some new WebExtension tabs would have no ETP setting in Fenix.

For https://github.com/mozilla-mobile/android-components/issues/9606
and -> https://github.com/mozilla-mobile/fenix/issues/12848
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks.
- [x] **Tests**: This PR includes thorough tests.
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md).
- [x] **Accessibility**: The code in this PR  does not include any user facing features.

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt [hunk 1]
```diff
@@ -239,16 +241,21 @@ class GeckoWebExtension(
...
+    override fun registerTabHandler(tabHandler: TabHandler, defaultSettings: Settings?) {
         val tabDelegate = object : GeckoNativeWebExtension.TabDelegate {
 
             override fun onNewTab(
                 ext: GeckoNativeWebExtension,
                 tabDetails: GeckoNativeWebExtension.CreateTabDetails
             ): GeckoResult<GeckoSession>? {
-                val geckoEngineSession = GeckoEngineSession(runtime, openGeckoSession = false)
+                val trackingPolicy = defaultSettings?.trackingProtectionPolicy
+                    ?: EngineSession.TrackingProtectionPolicy.strict()
```
Comment (csadilek) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt:252]: Why are we adding this fallback logic here? We shouldn't have to treat this session for extension tabs differently. 🤔 

If we have to we should handle this outside of the extension as it's unrelated code.

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt [hunk 2]
```diff
@@ -239,16 +241,21 @@ class GeckoWebExtension(
...
 
+                val geckoEngineSession = GeckoEngineSession(
+                    runtime,
+                    defaultSettings = DefaultSettings(trackingProtectionPolicy = trackingPolicy),
```
Comment (csadilek) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt:256]: Ideally we just pass the same instance of settings along here. See my comment above.

File: components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt [hunk 3]
```diff
@@ -244,15 +246,22 @@ class GeckoWebExtension(
     /**
      * See [WebExtension.registerTabHandler].
      */
-    override fun registerTabHandler(tabHandler: TabHandler) {
+    override fun registerTabHandler(tabHandler: TabHandler, defaultSettings: Settings?) {
 
         val tabDelegate = object : GeckoNativeWebExtension.TabDelegate {
 
+            val trackingPolicy = defaultSettings?.trackingProtectionPolicy
+                ?: EngineSession.TrackingProtectionPolicy.strict()
```
Comment (csadilek) [inline components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt:254]: See my comment above. Also here we have this outside `onNewTab` whereas in beta we have it inside, but ideally we shouldn't need this. The `GeckoWebExtension` should not need to know about tracking protection policies and just pass settings along.

**Review by csadilek — COMMENTED** [review]
I think the general shape of this is good! I have a comment about the TP defaults though which doesn't seem right to me to set here. Let's also make sure we smoke test some add-on setting screens with this in Fenix before landing.

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt (hunk 1 above)
Comment (Amejia481) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt:252]: Why are we creating an arbitrary setting for tracking protection? I think we we should pass the app settings.

**Comment by Amejia481** [comment]
@codrut-topliceanu I think we also need to pass the `defaultSettings`  when `onOpenOptionsPage` gets called, and all places where we are creating a new session.
https://github.com/mozilla-mobile/android-components/blob/acfb1c2b7ae55f1bdc02bc6c1348c1b81e4eb455/components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt#L272

**Comment by codrut-topliceanu** [comment]
Hey! Thanks for taking a look at this @csadilek & @Amejia481 🥇 

> @codrut-topliceanu I think we also need to pass the `defaultSettings` when `onOpenOptionsPage` gets called, and **all places where we are creating a new session**

@Amejia481 I don't disagree, but that would require A LOT of changes and this ticket is only to fix one problem with the ETP in Fenix. Regarding `onOpenOptionsPage`, sure I can add that there.

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt (hunk 2 above)
Comment (Amejia481) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt:254]: @codrut-topliceanu we are still creating a new settings object, let just pass the the reference (`defaultSettings`) that we get `registerTabHandler`.

**Comment by Amejia481** [comment]
> Hey! Thanks for taking a look at this @csadilek & @Amejia481 🥇
> 
> > @codrut-topliceanu I think we also need to pass the `defaultSettings` when `onOpenOptionsPage` gets called, and **all places where we are creating a new session**
> 
> @Amejia481 I don't disagree, but that would require A LOT of changes and this ticket is only to fix one problem with the ETP in Fenix. Regarding `onOpenOptionsPage`, sure I can add that there.

No problem when I said " **all places where we are creating a new session**" I meant in the context of this patch :)

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt (hunk 2 above)
Comment (Amejia481) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt:254]: Creating a new object, we are discarding all the previous and only keeping the `trackingProtectionPolicy`

Comment (codrut-topliceanu) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt:254]: I think I'm blind. Thank you!

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtension.kt [hunk 4]
```diff
@@ -128,7 +129,7 @@ abstract class WebExtension(
      * @param tabHandler the [TabHandler] to be invoked when the web extension
      * wants to open a new tab.
      */
-    abstract fun registerTabHandler(tabHandler: TabHandler)
+    abstract fun registerTabHandler(tabHandler: TabHandler, defaultSettings: Settings?)
```
Comment (Amejia481) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtension.kt:132]: nit: Let's add docs for the second parameter.

**Review by Amejia481 — APPROVED** [review]
Looks good! I found small nit about documentation, after squashing the commits this will be ready to 🚢 !

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt [hunk 5]
```diff
@@ -88,7 +88,8 @@ class GeckoEngine(
         }
 
         override fun onToggleActionPopup(extension: WebExtension, action: Action): EngineSession? {
-            return webExtensionDelegate?.onToggleActionPopup(extension, GeckoEngineSession(runtime), action)
+            return webExtensionDelegate?.onToggleActionPopup(extension, GeckoEngineSession(runtime,
+            defaultSettings = defaultSettings), action)
```
Comment (csadilek) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:92]: nit: intendation

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt [hunk 6]
```diff
@@ -262,7 +265,8 @@ class GeckoWebExtension(
                 ext.metaData.optionsPageUrl?.let { optionsPageUrl ->
                     tabHandler.onNewTab(
                         this@GeckoWebExtension,
-                        GeckoEngineSession(runtime),
+                        GeckoEngineSession(runtime,
+                            defaultSettings = defaultSettings),
```
Comment (csadilek) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/webextension/GeckoWebExtension.kt:269]: nit: formatting, could put this on the same line, or `)` on a new one.

File: components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt [hunk 7]
```diff
@@ -88,7 +88,8 @@ class GeckoEngine(
         }
 
         override fun onToggleActionPopup(extension: WebExtension, action: Action): EngineSession? {
-            return webExtensionDelegate?.onToggleActionPopup(extension, GeckoEngineSession(runtime), action)
+            return webExtensionDelegate?.onToggleActionPopup(extension, GeckoEngineSession(runtime,
+            defaultSettings = defaultSettings), action)
```
Comment (csadilek) [inline components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:92]: nit: indentation.

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt [hunk 8]
```diff
@@ -85,7 +85,8 @@ class GeckoEngine(
         }
 
         override fun onToggleActionPopup(extension: WebExtension, action: Action): EngineSession? {
-            return webExtensionDelegate?.onToggleActionPopup(extension, GeckoEngineSession(runtime), action)
+            return webExtensionDelegate?.onToggleActionPopup(extension, GeckoEngineSession(runtime,
+            defaultSettings = defaultSettings), action)
```
Comment (csadilek) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt:89]: nit: indentation

**Review by csadilek — APPROVED** [review]
Found some indentation issues, but otherwise ✅

**Comment by codrut-topliceanu** [comment]
Thanks for all your help ❤️ ! Think I got all the nits.
